text/0035-emergency-restarts.md

```diff
@@ -0,0 +1,116 @@
+
+In this CEP, we suggest a combination of options 2 and 3.
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The main thing required is to extend the `components::chainspec_loader::chainspec::UpgradePoint` structure to make it possible to specify arbitrary modifications to the global state.
+
+The best way of doing that is left as an unresolved question. We suggest adding the following two fields:
+
+```rust
+struct UpgradePoint {
+    // ... other fields
+    /// the paths to arbitrary WASM to be executed, in the specified order
+    upgrade_wasm_paths: Vec<String>,
+    /// the path to an `accounts.csv`-like file specifying the new set of validators
+    new_accounts: String,
+}
+```
+
+These would have to be translated into analogous fields in `execution_engine::core::engine_state::upgrade::UpgradeConfig` (but probably containing the data itself, ie. WASM binary bytes and the contents of `accounts.csv` instead of the paths).
+
+The execution engine would execute the specified WASM and update the accounts and the state of the auction contract as a part of `EngineState::commit_upgrade`.
+
+The data about activation points would have to be made available to the `BlockExecutor` and the `EraSupervisor`. The BlockExecutor would have to be aware of them in order to request the execution engine to commit the upgrades at the right moments. The EraSupervisor would need this data to properly determine switch blocks.
+
+`BlockHeader::switch_block` would now take an argument, which would be the list of heights of the upgrade points. A block would be a switch block if `era_end.is_some()` or if its height + 1 equals one of the heights of the upgrade points.
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+None at the moment.
+
+## Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+An alternative has been suggested in [CEP 6](https://github.com/CasperLabs/ceps/blob/master/text/0006-equivocation-catastrophes.md). It involved restarting the network as if from scratch, and using the mechanisms created for handling genesis to apply the necessary changes to the global state.
+
+While the genesis already does some of the things we need (like initializing the validator set), it is designed to do that on an empty global state. Modifying the logic there to make it apply to non-empty states as well is a complex task and doesn't gain us anything over just using the upgrades mechanism (which has been introduced exactly for applying arbitrary changes to the global state at specific heights).
+
+## Prior art
+
+[prior-art]: #prior-art
+
+None at the moment.
+
+## Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+- How should the arbitrary transforms to the global state be specified in the chainspec? Is the proposal from this CEP good enough?
```

Comment:
By "the proposal" you mean points 2 and 3 from *The structure of an upgrade point* ?